loading metadata from sidecar file

Johann150 opened the pull request:
This could resolve #6.
I am not sure how this should be tied into sending replies, since this would more or less be global state.
I am also not sure what should be put in the sidecar files. In theory these could contain the whole `<META>` part of the response, but it is probably undesirable to verbosely specify `text/gemini;lang=` for most cases. On the other hand this might be a nice feature for some other files. Or there could be two different formats where one lets the server decide the MIME type.

And this is once again synchronous code. I don't really see how this would be improved by making it async, but if preferred, I could try to rewrite it. Or maybe this should run in a separate thread using something like [notify](https://crates.io/crates/notify) to update the databases in the background.

I don't really know how to continue from here so any feedback would be greatly appreciated.

mbrubeck commented:
Thanks!  I’m wrestling with the same questions (like how to manage this global state, and how to allow both MIME and language overrides).  I'll let you know if I come up with any good answers.  :)

I'd definitely like to keep the code as simple as I reasonably can, but I'm willing to add complexity if it makes things better for users.

Johann150 commented:
One idea I hinted at which I think would be quite easy to implement would be to use "two formats", something like this

```text
index.gmi: ;charset=us-ascii;lang=en
dutch.gmi: ;lang=nl
german.gemini: text/gemini;lang=de
finnish: text/gemini;charset=ISO-8859-4;lang=fi
some.file: text/plain
```
Where the 1st and 2nd would guess the type of `text/gemini` based on the file extension.

The syntax of MIME types lends itself to this usage because of how parameters have to be separated from the rest of the type by a semicolon. But then also the question arises of how to handle or detect malformed lines, say I was to write `index.gmi: lang=en` (thus missing the semicolon), then agate would be blissfully unaware and serve that up as gemini response `20 lang=en\r\n...`. Depends on if this would be okay to just put in the responsibility of the administrator that they have to get this right.
The only way around that would be to try to parse every line that does not start with a semicolon as a MIME type and if that fails then... maybe throw it out like other invalid lines. But first trying to parse every line of that file could potentially be quite a task.

Johann150 commented:
I implemented the system I described above, and also fixed the detection of when the sidecar file is outdated.

The "global" state is handled by using an `Arc<RwLock<_>>` that is passed along as a function parameter.
This could be simplified by using something like a `ClientHandler` struct which could store `stream` and this new value. Doing this would reduce the number of parameters that need to be passed along, but since its still okay I don't think this is necessary now.

The directory listing response is without the language parameter again, as I don't think it is of much value for a filename listing. But with a little bit of extra code I could add it back in.

mbrubeck commented:
Thanks again! This is very useful.

I'm busy with work this week, so I may not have time to review this completely for a little while.  I'm also not 100% sure that I want to go this direction, so I hope if it's okay if I might end up choosing a different approach.

Ylhp commented:
I didn't have a good rust level to be sure that works, but at least, that means little as adding, and not stupidly big : good point. And in plus, that let the mime format to be controlled, not only the lang header. Thanks for this !

I keap an ear here to stay tunned ! :)

Johann150 commented:
Just an idea: Also allow entries in the sidecar file for nonexistent files, for things like status codes `30 REDIRECT - TEMPORARY`, `31 REDIRECT - PERMANENT` and  `52 GONE`. But this would also require a bit more complications with the format which should be resolvable unambiguously by looking at the first character:
- number &rarr; must be full header line.
- semicolon &rarr; must be status `20` with mime type and append the remaining line as parameters.
- else must be status `20` and remaining line as META.
- (if no matching line, same behaviour as before this pull request)

Not sure if this is a good idea; supporting this seems nice and since the file has to be parsed anyway, the difference shouldn't be too big. I would appreciate feedback!

mbrubeck commented on README.md:
```diff
@@ -43,6 +43,16 @@ All of the command-line arguments are optional.  Run `agate --help` to see the d
 
 When a client requests the URL `gemini://example.com/foo/bar`, Agate will respond with the file at `path/to/content/foo/bar`. If any segment of the requested path starts with a dot, agate will respond with a status code 52, even if the file does not exist (this behaviour can be disabled with `--serve-secret`). If there is a directory at that path, Agate will look for a file named `index.gmi` insi ...
 
+Agate will look for a file called `.mime` in the same directory as the file currently being served. If this file exists and has an entry for the current file, the respective data will be used to formulate the response header.
```
Maybe there should be a section heading to introduce this part.

mbrubeck commented on README.md:
````diff
@@ -43,6 +43,16 @@ All of the command-line arguments are optional.  Run `agate --help` to see the d
+<filename>:<metadata>
+```
+
+Where `<filename>` is just a filename (not a path) of a file in the same directory, and `<metadata>` is the MIME type to be stored. If `<metadata>` starts with a semicolon, agate will use the usual mechanism to determine the mime type of the file and append the specified parameters.
````
`FullHeader` lines are not yet documented here.

Johann150 replied:
I was still working on the updated documentation, maybe you can take a quick look at it again?

mbrubeck commented on src/main.rs:
```diff
@@ -142,12 +179,13 @@ fn acceptor() -> Result<TlsAcceptor> {
 struct RequestHandle {
     stream: TlsStream<TcpStream>,
     log_line: String,
+    metadata: Arc<RwLock<FileOptions>>,
```
Maybe this should be a `Mutex` instead of `RwLock` because we never use `.read()` access.

Johann150 replied:
Ah, yes I was thinking about this but forgot to actually do it. Thanks for reminding!

mbrubeck reviewed (approved):
This looks great.  I especially appreciate how well-commented it is.  Some minor suggestions below.
